File: leetcodes_devrod/buy_and_sell/buy_and_sell.py
def buy_and_sell_with_max_profit(prices: list[int]) -> int:
    if not prices or len(prices) <2:
        return 0
    min_price_so_far = prices

    max_profit_so_far = 0

    for i in range(1, len(prices)):
        current_price = prices[i]
        potential_profit = current_price - min_price_so_far
        max_profit_so_far = max(max_profit_so_far, potential_profit)
        min_price_so_far = min(min_price_so_far, current_price)
        return max_profit_so_far

File: leetcodes_devrod/arrays/buy_and_sell/buy_and_sell.ts
export const buyAndSellWithMaxProfit = (prices: number[]): number => {
    
    let minPrice = Infinity;
    let maxProfit = 0;

    for(let i = 0; i < prices.length; i++){
        if(prices[i] < minPrice){
            minPrice = prices[i];
        } else if(prices[i] - minPrice > maxProfit){
            maxProfit = prices[i] - minPrice;
        }
        console.log(`Day ${i + 1}: Price = ${prices[i]}, Min Price = ${minPrice}, Max Profit = ${maxProfit}`);

    }
    return maxProfit;
}